#VRML_SIM R2021a utf8
WorldInfo {
  basicTimeStep 16
  coordinateSystem "NUE"
}
Viewpoint {
  orientation -0.5231075656343174 0.8285928552018662 0.19948021226316554 0.8626443644312318
  position 0.862902715562721 0.6864253633365853 0.9155411528425393
}
Floor {
}
TexturedBackground {
}
TexturedBackgroundLight {
}
DEF FLY Robot {
  translation 0 0.2 0
  children [
    DEF HEAD TouchSensor {
      translation 0 -0.01 0.11
      rotation 1 0 0 -0.1308996938995747
      children [
        DEF HEAD_SHAPE Shape {
          appearance DEF BODY_APPEARANCE PBRAppearance {
            baseColor 0.666667 0.333333 0
            roughness 1
            metalness 0
          }
          geometry Capsule {
            height 0.02
            radius 0.05
          }
        }
      ]
      name "head"
      boundingObject USE HEAD_SHAPE
      physics Physics {
      }
    }
    DEF THORAX TouchSensor {
      children [
        DEF THORAX_SHAPE Shape {
          appearance USE BODY_APPEARANCE
          geometry Sphere {
            radius 0.07
            subdivision 2
          }
        }
      ]
      name "thorax"
      boundingObject USE THORAX_SHAPE
      physics Physics {
      }
    }
    DEF ABDOMEN TouchSensor {
      translation 0 -0.02 -0.14
      rotation 1 0 0 -1.7017003061004252
      children [
        DEF ABDOMEN_SHAPE Shape {
          appearance USE BODY_APPEARANCE
          geometry Capsule {
            height 0.08
            radius 0.06
          }
        }
      ]
      name "abdomen"
      boundingObject USE ABDOMEN_SHAPE
      physics Physics {
      }
    }
    DEF LF_TC BallJoint {
      jointParameters BallJointParameters {
        anchor 0.02 -0.04 0.05
      }
      jointParameters2 JointParameters {
        axis 0 -1 0
      }
      jointParameters3 JointParameters {
      }
      device [
        RotationalMotor {
          name "lf_tc_pro/re"
        }
      ]
      device2 [
        RotationalMotor {
          name "lf_tc_rot"
          minPosition -1.5707963267948966
          maxPosition 1.5707963267948966
        }
      ]
      device3 [
        RotationalMotor {
          name "lf_tc_abd/add"
        }
      ]
      endPoint DEF LF_COXA TouchSensor {
        translation 0.019999999825634 -0.04999999877768204 0.04999505567422416
        rotation 0.0009254430140561519 -0.9999995410325083 -0.0002479717739590616 0.5235892217062273
        children [
          DEF COXA_SHAPE Shape {
            appearance USE BODY_APPEARANCE
            geometry Capsule {
              height 0.02
              radius 0.01
            }
          }
          DEF LF_CT HingeJoint {
            jointParameters HingeJointParameters {
              position 0.4
              axis 0 0 1
              anchor 0 -0.01 0
            }
            device [
              RotationalMotor {
                name "lf_ct_flex/ext"
              }
            ]
            endPoint DEF LF_FEMUR TouchSensor {
              translation 0.03684243976011533 0.005576733692345999 0
              rotation 0 0 1 1.9707996938995735
              children [
                DEF FEMUR_SHAPE Shape {
                  appearance USE BODY_APPEARANCE
                  geometry Capsule {
                    height 0.08
                    radius 0.01
                  }
                }
                DEF LF_FT HingeJoint {
                  jointParameters HingeJointParameters {
                    position 0.2
                    axis 0 0 1
                    anchor 0 -0.04 0
                  }
                  device [
                    RotationalMotor {
                      name "lf_ft_flex/ext"
                    }
                  ]
                  endPoint DEF LF_TIBIA TouchSensor {
                    translation -0.039202663113650255 -0.0479467732318026 0
                    rotation 0 0 -1 1.3707996938995706
                    children [
                      DEF TIBIA_SHAPE Shape {
                        appearance USE BODY_APPEARANCE
                        geometry Capsule {
                          height 0.08
                          radius 0.008
                        }
                      }
                      DEF LF_TT HingeJoint {
                        jointParameters HingeJointParameters {
                          position 0.6
                          axis 0 0 1
                          anchor 0 -0.04 0
                        }
                        device [
                          RotationalMotor {
                            name "lf_tt_flex/ext"
                          }
                        ]
                        endPoint DEF LF_TARSUS TouchSensor {
                          translation 0.016939274201851047 -0.06476006844729036 0
                          rotation 0 0 1 0.6000000000000003
                          children [
                            DEF TARSUS_SHAPE Shape {
                              appearance USE BODY_APPEARANCE
                              geometry Capsule {
                                height 0.06
                                radius 0.005
                              }
                            }
                            DEF LF_CLAW TouchSensor {
                              translation 0 -0.03 0
                              children [
                                DEF CLAW_SHAPE Shape {
                                  appearance USE BODY_APPEARANCE
                                  geometry Sphere {
                                    radius 0.005
                                    subdivision 2
                                  }
                                }
                              ]
                              name "lf_claw"
                              boundingObject USE CLAW_SHAPE
                              physics DEF LEG_PHYSICS Physics {
                                density -1
                                mass 1
                              }
                            }
                          ]
                          name "lf_tarsus"
                          boundingObject USE TARSUS_SHAPE
                          physics USE LEG_PHYSICS
                        }
                      }
                    ]
                    name "lf_tibia"
                    boundingObject USE TIBIA_SHAPE
                    physics USE LEG_PHYSICS
                  }
                }
              ]
              boundingObject USE FEMUR_SHAPE
              physics USE LEG_PHYSICS
            }
          }
        ]
        name "lf_coxa"
        boundingObject USE COXA_SHAPE
        physics USE LEG_PHYSICS
      }
    }
    DEF LM_TC BallJoint {
      jointParameters BallJointParameters {
        anchor 0.02 -0.06 0
      }
      jointParameters2 JointParameters {
        axis 0 -1 0
      }
      jointParameters3 JointParameters {
      }
      device [
        RotationalMotor {
          name "lm_tc_pro/re"
        }
      ]
      device2 [
        RotationalMotor {
          name "lm_tc_rot"
          minPosition -1.5707963267948966
          maxPosition 1.5707963267948966
        }
      ]
      device3 [
        RotationalMotor {
          name "lm_tc_abd/add"
        }
      ]
      endPoint DEF LM_COXA TouchSensor {
        translation 0.01999999806056044 -0.06997322855646432 -0.000731240152426596
        rotation 0.23376748143941511 -0.46079578373164304 0.8561658778044202 0.597080801701449
        children [
          USE COXA_SHAPE
          DEF LM_CT HingeJoint {
            jointParameters HingeJointParameters {
              position 0.1
              axis 0 0 1
              anchor 0 -0.01 0
            }
            device [
              RotationalMotor {
                name "lm_ct_flex/ext"
              }
            ]
            endPoint DEF LM_FEMUR TouchSensor {
              translation 0.039800166611121374 -0.006006663334126841 0
              rotation 0 0 1 1.6707996938995746
              children [
                USE FEMUR_SHAPE
                DEF LM_FT HingeJoint {
                  jointParameters HingeJointParameters {
                    position 0.1
                    axis 0 0 1
                    anchor 0 -0.04 0
                  }
                  device [
                    RotationalMotor {
                      name "lm_ft_flex/ext"
                    }
                  ]
                  endPoint DEF LM_TIBIA TouchSensor {
                    translation -0.0398001666111214 -0.04399333666587317 0
                    rotation 0 0 -1 1.4707996938995747
                    children [
                      USE TIBIA_SHAPE
                      DEF LM_TT HingeJoint {
                        jointParameters HingeJointParameters {
                          position 0.5
                          axis 0 0 1
                          anchor 0 -0.04 0
                        }
                        device [
                          RotationalMotor {
                            name "lm_tt_flex/ext"
                          }
                        ]
                        endPoint DEF LM_TARSUS TouchSensor {
                          translation 0.01438276615812609 -0.06632747685671118 0
                          rotation 0 0 1 0.5000000000000004
                          children [
                            USE TARSUS_SHAPE
                            DEF LM_CLAW TouchSensor {
                              translation 0 -0.03 0
                              children [
                                USE CLAW_SHAPE
                              ]
                              name "lm_claw"
                              boundingObject USE CLAW_SHAPE
                              physics USE LEG_PHYSICS
                            }
                          ]
                          name "lm_tarsus"
                          boundingObject USE TARSUS_SHAPE
                          physics USE LEG_PHYSICS
                        }
                      }
                    ]
                    name "lm_tibia"
                    boundingObject USE TIBIA_SHAPE
                    physics USE LEG_PHYSICS
                  }
                }
              ]
              boundingObject USE FEMUR_SHAPE
              physics USE LEG_PHYSICS
            }
          }
        ]
        name "lm_coxa"
        boundingObject USE COXA_SHAPE
        physics USE LEG_PHYSICS
      }
    }
    DEF LR_TC BallJoint {
      jointParameters BallJointParameters {
        anchor 0.02 -0.04 -0.05
      }
      jointParameters2 JointParameters {
        axis 0 -1 0
      }
      jointParameters3 JointParameters {
      }
      device [
        RotationalMotor {
          name "lr_tc_pro/re"
        }
      ]
      device2 [
        RotationalMotor {
          name "lr_tc_rot"
          minPosition -1.5707963267948966
          maxPosition 1.5707963267948966
        }
      ]
      device3 [
        RotationalMotor {
          name "lr_tc_abd/add"
        }
      ]
      endPoint DEF LR_COXA TouchSensor {
        translation 0.020000005657068082 -0.04987600037181521 -0.051569909760423756
        rotation 0.8610910966652042 -0.23132979851772498 0.45277880644047896 1.0690987444018925
        children [
          USE COXA_SHAPE
          DEF LR_CT HingeJoint {
            jointParameters HingeJointParameters {
              position -0.4
              axis 0 0 1
              anchor 0 -0.01 0
            }
            device [
              RotationalMotor {
                name "lr_ct_flex/ext"
              }
            ]
            endPoint DEF LR_FEMUR TouchSensor {
              translation 0.03684243976011537 -0.025576733692346004 0
              rotation 0 0 1 1.1707996938995748
              children [
                USE FEMUR_SHAPE
                DEF LR_FT HingeJoint {
                  jointParameters HingeJointParameters {
                    position 0.5
                    axis 0 0 1
                    anchor 0 -0.04 0
                  }
                  device [
                    RotationalMotor {
                      name "lr_ft_flex/ext"
                    }
                  ]
                  endPoint DEF LR_TIBIA TouchSensor {
                    translation -0.03510330247561486 -0.05917702154416811 0
                    rotation 0 0 -1 1.0707996938995759
                    children [
                      USE TIBIA_SHAPE
                      DEF LR_TT HingeJoint {
                        jointParameters HingeJointParameters {
                          position 0.6
                          axis 0 0 1
                          anchor 0 -0.04 0
                        }
                        device [
                          RotationalMotor {
                            name "lr_tt_flex/ext"
                          }
                        ]
                        endPoint DEF LR_TARSUS TouchSensor {
                          translation 0.016939274201851026 -0.06476006844729028 0
                          rotation 0 0 1 0.6000000000000003
                          children [
                            USE TARSUS_SHAPE
                            DEF LR_CLAW TouchSensor {
                              translation 0 -0.03 0
                              children [
                                USE CLAW_SHAPE
                              ]
                              name "lr_claw"
                              boundingObject USE CLAW_SHAPE
                              physics USE LEG_PHYSICS
                            }
                          ]
                          name "lr_tarsus"
                          boundingObject USE TARSUS_SHAPE
                          physics USE LEG_PHYSICS
                        }
                      }
                    ]
                    name "lr_tibia"
                    boundingObject USE TIBIA_SHAPE
                    physics USE LEG_PHYSICS
                  }
                }
              ]
              boundingObject USE FEMUR_SHAPE
              physics USE LEG_PHYSICS
            }
          }
        ]
        name "lr_coxa"
        boundingObject USE COXA_SHAPE
        physics USE LEG_PHYSICS
      }
    }
    DEF RF_TC BallJoint {
      jointParameters BallJointParameters {
        anchor -0.02 -0.04 0.05
      }
      jointParameters2 JointParameters {
        axis 0 1 0
      }
      jointParameters3 JointParameters {
        axis 0 0 -1
      }
      device [
        RotationalMotor {
          name "rf_tc_pro/re"
        }
      ]
      device2 [
        RotationalMotor {
          name "rf_tc_rot"
          minPosition -1.5707963267948966
          maxPosition 1.5707963267948966
        }
      ]
      device3 [
        RotationalMotor {
          name "rf_tc_abd/add"
        }
      ]
      endPoint DEF RF_COXA TouchSensor {
        translation -0.01999999992165968 -0.049999968063315 0.05002527316125795
        rotation -0.0003392634660045601 -0.9999991408301727 0.0012661908295375937 2.618016334536957
        children [
          USE COXA_SHAPE
          DEF RF_CT HingeJoint {
            jointParameters HingeJointParameters {
              position 0.4
              axis 0 0 1
              anchor 0 -0.01 0
            }
            device [
              RotationalMotor {
                name "rf_ct_flex/ext"
              }
            ]
            endPoint DEF RF_FEMUR TouchSensor {
              translation 0.03684243976011537 0.0055767336923460056 0
              rotation 0 0 1 1.9707996938995735
              children [
                USE FEMUR_SHAPE
                DEF RF_FT HingeJoint {
                  jointParameters HingeJointParameters {
                    position 0.2
                    axis 0 0 1
                    anchor 0 -0.04 0
                  }
                  device [
                    RotationalMotor {
                      name "rf_ft_flex/ext"
                    }
                  ]
                  endPoint DEF RF_TIBIA TouchSensor {
                    translation -0.039202663113650234 -0.04794677323180257 0
                    rotation 0 0 -1 1.3707996938995735
                    children [
                      USE TIBIA_SHAPE
                      DEF RF_TT HingeJoint {
                        jointParameters HingeJointParameters {
                          position 0.6
                          axis 0 0 1
                          anchor 0 -0.04 0
                        }
                        device [
                          RotationalMotor {
                            name "lf_tt_flex/ext"
                          }
                        ]
                        endPoint DEF RF_TARSUS TouchSensor {
                          translation 0.016939274201851026 -0.06476006844729028 0
                          rotation 0 0 1 0.6000000000000003
                          children [
                            USE TARSUS_SHAPE
                            DEF RF_CLAW TouchSensor {
                              translation 0 -0.03 0
                              children [
                                USE CLAW_SHAPE
                              ]
                              name "rf_claw"
                              boundingObject USE CLAW_SHAPE
                              physics USE LEG_PHYSICS
                            }
                          ]
                          name "rf_tarsus"
                          boundingObject USE TARSUS_SHAPE
                          physics USE LEG_PHYSICS
                        }
                      }
                    ]
                    name "rf_tibia"
                    boundingObject USE TIBIA_SHAPE
                    physics USE LEG_PHYSICS
                  }
                }
              ]
              boundingObject USE FEMUR_SHAPE
              physics USE LEG_PHYSICS
            }
          }
        ]
        name "rf_coxa"
        boundingObject USE COXA_SHAPE
        physics USE LEG_PHYSICS
      }
    }
    DEF RM_TC BallJoint {
      jointParameters BallJointParameters {
        anchor -0.02 -0.06 0
      }
      jointParameters2 JointParameters {
        axis 0 1 0
      }
      jointParameters3 JointParameters {
        axis 0 0 -1
      }
      device [
        RotationalMotor {
          name "rm_tc_pro/re"
        }
      ]
      device2 [
        RotationalMotor {
          name "rm_tc_rot"
          minPosition -1.5707963267948966
          maxPosition 1.5707963267948966
        }
      ]
      device3 [
        RotationalMotor {
          name "rm_tc_abd/add"
        }
      ]
      endPoint DEF RM_COXA TouchSensor {
        translation -0.02000049585891499 -0.06997251925643018 -0.0007408504803626957
        rotation -0.2538537561833751 -0.9672425946991559 0.00018295145742211607 2.855533228143019
        children [
          USE COXA_SHAPE
          DEF RM_CT HingeJoint {
            jointParameters HingeJointParameters {
              position 0.1
              axis 0 0 1
              anchor 0 -0.01 0
            }
            device [
              RotationalMotor {
                name "rm_ct_flex/ext"
              }
            ]
            endPoint DEF RM_FEMUR TouchSensor {
              translation 0.039800166611121374 -0.006006663334126841 0
              rotation 0 0 1 1.6707996938995746
              children [
                USE FEMUR_SHAPE
                DEF RM_FT HingeJoint {
                  jointParameters HingeJointParameters {
                    position 0.1
                    axis 0 0 1
                    anchor 0 -0.04 0
                  }
                  device [
                    RotationalMotor {
                      name "rm_ft_flex/ext"
                    }
                  ]
                  endPoint DEF RM_TIBIA TouchSensor {
                    translation -0.0398001666111214 -0.04399333666587317 0
                    rotation 0 0 -1 1.4707996938995747
                    children [
                      USE TIBIA_SHAPE
                      DEF RM_TT HingeJoint {
                        jointParameters HingeJointParameters {
                          position 0.5
                          axis 0 0 1
                          anchor 0 -0.04 0
                        }
                        device [
                          RotationalMotor {
                            name "lf_tt_flex/ext"
                          }
                        ]
                        endPoint DEF RM_TARSUS TouchSensor {
                          translation 0.01438276615812609 -0.06632747685671118 0
                          rotation 0 0 1 0.5000000000000004
                          children [
                            USE TARSUS_SHAPE
                            DEF RM_CLAW TouchSensor {
                              translation 0 -0.03 0
                              children [
                                USE CLAW_SHAPE
                              ]
                              name "rm_claw"
                              boundingObject USE CLAW_SHAPE
                              physics USE LEG_PHYSICS
                            }
                          ]
                          name "rm_tarsus"
                          boundingObject USE TARSUS_SHAPE
                          physics USE LEG_PHYSICS
                        }
                      }
                    ]
                    name "rm_tibia"
                    boundingObject USE TIBIA_SHAPE
                    physics USE LEG_PHYSICS
                  }
                }
              ]
              boundingObject USE FEMUR_SHAPE
              physics USE LEG_PHYSICS
            }
          }
        ]
        name "rm_coxa"
        boundingObject USE COXA_SHAPE
        physics USE LEG_PHYSICS
      }
    }
    DEF RR_TC BallJoint {
      jointParameters BallJointParameters {
        anchor -0.02 -0.04 -0.05
      }
      jointParameters2 JointParameters {
        axis 0 1 0
      }
      jointParameters3 JointParameters {
        axis 0 0 -1
      }
      device [
        RotationalMotor {
          name "rr_tc_pro/re"
        }
      ]
      device2 [
        RotationalMotor {
          name "rr_tc_rot"
          minPosition -1.5707963267948966
          maxPosition 1.5707963267948966
        }
      ]
      device3 [
        RotationalMotor {
          name "rr_tc_abd/add"
        }
      ]
      endPoint DEF RR_COXA TouchSensor {
        translation -0.020000002627955708 -0.049872401730184576 -0.05159238314417164
        rotation -0.23205678416351744 -0.866055361302756 -0.4428292674185385 2.9054071811679827
        children [
          USE COXA_SHAPE
          DEF RR_CT HingeJoint {
            jointParameters HingeJointParameters {
              position -0.4
              axis 0 0 1
              anchor 0 -0.01 0
            }
            device [
              RotationalMotor {
                name "rr_ct_flex/ext"
              }
            ]
            endPoint DEF RR_FEMUR TouchSensor {
              translation 0.03684243976011537 -0.025576733692346004 0
              rotation 0 0 1 1.1707996938995748
              children [
                USE FEMUR_SHAPE
                DEF RR_FT HingeJoint {
                  jointParameters HingeJointParameters {
                    position 0.5
                    axis 0 0 1
                    anchor 0 -0.04 0
                  }
                  device [
                    RotationalMotor {
                      name "rr_ft_flex/ext"
                    }
                  ]
                  endPoint DEF RR_TIBIA TouchSensor {
                    translation -0.03510330247561486 -0.05917702154416811 0
                    rotation 0 0 -1 1.0707996938995759
                    children [
                      USE TIBIA_SHAPE
                      DEF RR_TT HingeJoint {
                        jointParameters HingeJointParameters {
                          position 0.6
                          axis 0 0 1
                          anchor 0 -0.04 0
                        }
                        device [
                          RotationalMotor {
                            name "lf_tt_flex/ext"
                          }
                        ]
                        endPoint DEF RR_TARSUS TouchSensor {
                          translation 0.016939274201851026 -0.06476006844729028 0
                          rotation 0 0 1 0.6000000000000003
                          children [
                            USE TARSUS_SHAPE
                            DEF RR_CLAW TouchSensor {
                              translation 0 -0.03 0
                              children [
                                USE CLAW_SHAPE
                              ]
                              name "rr_claw"
                              boundingObject USE CLAW_SHAPE
                              physics USE LEG_PHYSICS
                            }
                          ]
                          name "rr_tarsus"
                          boundingObject USE TARSUS_SHAPE
                          physics USE LEG_PHYSICS
                        }
                      }
                    ]
                    name "rr_tibia"
                    boundingObject USE TIBIA_SHAPE
                    physics USE LEG_PHYSICS
                  }
                }
              ]
              boundingObject USE FEMUR_SHAPE
              physics USE LEG_PHYSICS
            }
          }
        ]
        name "rr_coxa"
        boundingObject USE COXA_SHAPE
        physics USE LEG_PHYSICS
      }
    }
  ]
  boundingObject USE THORAX_SHAPE
  physics Physics {
  }
}
